src/messaging/ReliableMessageContext.cpp

```diff
@@ -0,0 +1,493 @@
+/*
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements the CHIP reliable message protocol.
+ *
+ */
+
+#include <inttypes.h>
+
+#include <messaging/ReliableMessageContext.h>
+
+#include <core/CHIPEncoding.h>
+#include <messaging/ErrorCategory.h>
+#include <messaging/Flags.h>
+#include <messaging/ReliableMessageManager.h>
+#include <protocols/CHIPProtocols.h>
+#include <protocols/common/CommonProtocol.h>
+#include <support/CodeUtils.h>
+
+namespace chip {
+namespace messaging {
+
+void ReliableMessageContextDeletor::Release(ReliableMessageContext * obj)
+{
+    ReliableMessageManager::GetManager().FreeContext(obj);
+}
+
+ReliableMessageContext::ReliableMessageContext() :
+    mConfig(gDefaultReliableMessageProtocolConfig), mNextAckTimeTick(0), mThrottleTimeoutTick(0), mPendingPeerAckId(0),
+    mDelegate(nullptr)
+{}
+
+/**
+ *  Determine whether there is already an acknowledgment pending to be sent
+ *  to the peer on this exchange.
+ *
+ */
+bool ReliableMessageContext::IsAckPending() const
+{
+    return mFlags.Has(Flags::kFlagAckPending);
+}
+
+/**
+ *  Determine whether peer requested acknowledgment for at least one message
+ *  on this exchange.
+ *
+ *  @return Returns 'true' if acknowledgment requested, else 'false'.
+ */
+bool ReliableMessageContext::HasPeerRequestedAck() const
+{
+    return mFlags.Has(Flags::kFlagPeerRequestedAck);
+}
+
+/**
+ *  Determine whether at least one message has been received
+ *  on this exchange from peer.
+ *
+ *  @return Returns 'true' if message received, else 'false'.
+ */
+bool ReliableMessageContext::HasRcvdMsgFromPeer() const
+{
+    return mFlags.Has(Flags::kFlagMsgRcvdFromPeer);
+}
+
+/**
+ *  Set if a message has been received from the peer
+ *  on this exchange.
+ *
+ *  @param[in]  inMsgRcvdFromPeer  A Boolean indicating whether (true) or not
+ *                                 (false) a message has been received
+ *                                 from the peer on this exchange context.
+ *
+ */
+void ReliableMessageContext::SetMsgRcvdFromPeer(bool inMsgRcvdFromPeer)
+{
+    if (inMsgRcvdFromPeer)
```

Comment:
nit: I think BitFlags has a Set(flag, bool) member, so the functions below could be like:

`mFlags.Set(Flags::kFlagMsgRcvdFromPeer, inMsgRecvdFromPeer);`

Reply:
done